Make Kokkos be usable in multiple CMake languages

JBludau opened the pull request:
This addition allows to use Kokkos in multiple CMake languages independently of Kokkos itself being built with CUDA or HIP (via `Kokkos_ENABLE_COMPILE_AS_CMAKE_LANGUAGE`). This allows downstream projects to use the same Kokkos in a lib that identifies as CUDA language to CMake and another lib that identifies as CXX to CMake. 

This is only active if `Kokkos_ENABLE_MULTIPLE_CMAKE_LANGUAGES=ON`. If that option is not given, the `Kokkos_COMPILE_LANGUAGE` will be `CXX` or `CUDA/HIP` depending on how Kokkos was built.

JBludau committed:
add cxx to language builds regardless

Signed-off-by: Jakob Bludau <jakob.bludau@tum.de>
Signed-off-by: Jakob Bludau <bludauj@ornl.gov>

JBludau commented on cmake/kokkos_tribits.cmake:
```diff
@@ -248,18 +248,18 @@ function(KOKKOS_SET_LIBRARY_PROPERTIES LIBRARY_NAME)
   if(KOKKOS_ENABLE_CUDA)
     target_compile_options(
-      ${LIBRARY_NAME} PUBLIC $<$<COMPILE_LANGUAGE:${KOKKOS_COMPILE_LANGUAGE}>:${KOKKOS_CUDA_OPTIONS}>
+      ${LIBRARY_NAME} PUBLIC $<$<COMPILE_LANGUAGE:${KOKKOS_COMPILE_LANGUAGE},CXX>:${KOKKOS_CUDA_OPTIONS}>
```
@cedricchevalier19 we can not spell out `CUDA` here since this would be a behavior change for builds without `Kokkos_ENABLE_COMPILE_AS_CMAKE_LANGUAGE`

JBludau committed:
add option Kokkos_ENABLE_MULTIPLE_CMAKE_LANGUAGES

Signed-off-by: Jakob Bludau <jakob.bludau@tum.de>
Signed-off-by: Jakob Bludau <bludauj@ornl.gov>

JBludau committed:
use new option to set languages

Signed-off-by: Jakob Bludau <jakob.bludau@tum.de>
Signed-off-by: Jakob Bludau <bludauj@ornl.gov>

JBludau committed:
cmake-format

Signed-off-by: Jakob Bludau <jakob.bludau@tum.de>
Signed-off-by: Jakob Bludau <bludauj@ornl.gov>

masterleinad commented:
Can you elaborate a little on the motivation? Which use cases does this enable?

JBludau commented:
Usecase: An application downstream uses some libraries that use CUDA code and other libs that use Kokkos. With Kokkos compiled as CXX language, it can't be used in the CUDA code, without setting the language of the respective file to CUDA in CMake. The same holds vice versa.

Kokkos is an inherently compiler-specific instantiation of our machine model for a respective backend. But this is independent of which language we identify as for CMake. With the proposed change, the same installed Kokkos can be used in both CUDA/HIP files or CXX files. It basically identifies as both a CXX and CUDA/HIP library for CMake

masterleinad commented:
What's the reason that we need an extra CMake option instead of always allowing multiple languages (if `Kokkos_ENABLE_COMPILE_AS_CMAKE_LANGUAGE` is enabled)?

JBludau commented:
> What's the reason that we need an extra CMake option instead of always allowing multiple languages (if `Kokkos_ENABLE_COMPILE_AS_CMAKE_LANGUAGE` is enabled)?

Previously we always only had one language that we identified as (CXX normally, and CUDA/HIP with the language option). This also means, that we only added flags to source files that were compiled in the respective languages. If we just always do a multilanguage install now, we might break someone that is relying on Kokkos only identifying as one language.

Rombur commented on cmake/kokkos_tribits.cmake:
```diff
@@ -223,10 +223,16 @@ endmacro()
 function(KOKKOS_SET_LIBRARY_PROPERTIES LIBRARY_NAME)
   cmake_parse_arguments(PARSE "PLAIN_STYLE" "" "" ${ARGN})
 
+  if(Kokkos_ENABLE_MULTIPLE_CMAKE_LANGUAGES)
```
Should we check that `COMPILE_AS_CMAKE_LANGUAGE` is ON?

JBludau replied:
I changed the condition below so `MULITPLE_CMAKE_LANGUAGES` is independent of `COMPILE_AS_CMAKE_LANGUAGE`

JBludau committed:
make multilang independent of compile_as_cmake_language

Signed-off-by: Jakob Bludau <jakob.bludau@tum.de>
Signed-off-by: Jakob Bludau <bludauj@ornl.gov>

JBludau committed:
added missing endif

Signed-off-by: Jakob Bludau <jakob.bludau@tum.de>
Signed-off-by: Jakob Bludau <bludauj@ornl.gov>

Rombur reviewed (approved):
I think it's good but it will need clear documentation.

masterleinad commented:
I'd appreciate if we added some kind of test clarifying the use case.

JBludau committed:
enable launch compiler for multilang cuda builds

Signed-off-by: Jakob Bludau <jakob.bludau@tum.de>
Signed-off-by: Jakob Bludau <bludauj@ornl.gov>

JBludau committed:
add a multilanguage example

Signed-off-by: Jakob Bludau <jakob.bludau@tum.de>
Signed-off-by: Jakob Bludau <bludauj@ornl.gov>

JBludau committed:
build multilanguage example in jenkins

Signed-off-by: Jakob Bludau <jakob.bludau@tum.de>
Signed-off-by: Jakob Bludau <bludauj@ornl.gov>

JBludau commented:
> I'd appreciate if we added some kind of test clarifying the use case.

I added an example that uses the same kokkos installation in multiple languages

JBludau committed:
apply cmake format

Signed-off-by: Jakob Bludau <jakob.bludau@tum.de>
Signed-off-by: Jakob Bludau <bludauj@ornl.gov>

JBludau committed:
forgot to commit one format change

Signed-off-by: Jakob Bludau <jakob.bludau@tum.de>
Signed-off-by: Jakob Bludau <bludauj@ornl.gov>

masterleinad commented on .jenkins/continuous.groovy:
```diff
@@ -166,6 +167,14 @@ pipeline {
+                                -DCMAKE_CXX_FLAGS=-Werror \
+                                -DCMAKE_CXX_STANDARD=17 \
+                              .. && \
+                              make -j8 && ctest --verbose && \
```
```
cd ../build_cmake_installed_multilanguage

/var/jenkins/workspace/Kokkos_PR-8167@tmp/durable-87daad5e/script.sh.copy: 1: cd: can't cd to ../build_cmake_installed_multilanguage
```
```suggestion
                              cd ../example/build_cmake_installed_multilanguage && \
                              rm -rf build && mkdir -p build && cd build && \
                              cmake \
                                -DCMAKE_CXX_COMPILER=g++-8 \
                                -DCMAKE_CXX_FLAGS=-Werror \
                                -DCMAKE_CXX_STANDARD=17 \
                              .. && \
                              make -j8 && ctest --verbose && \
```

JBludau committed:
Update .jenkins/continuous.groovy

Co-authored-by: Daniel Arndt <arndtd@ornl.gov>
Signed-off-by: Jakob Bludau <bludauj@ornl.gov>

JBludau commented:
Retest this please

JBludau committed:
correct path to example

Signed-off-by: Jakob Bludau <jakob.bludau@tum.de>
Signed-off-by: Jakob Bludau <bludauj@ornl.gov>

JBludau commented:
This was creating problems with Cuda 11, since the CI is Cuda 12 now, lets see if it works

JBludau committed:
remove old cxx17

Signed-off-by: Jakob Bludau <bludauj@ornl.gov>

JBludau commented:
hmm ... it reports missing includes now. Did not see that before

JBludau committed:
bump cmake to 3.25.3 for multilanguage feature

Signed-off-by: Jakob Bludau <bludauj@ornl.gov>

JBludau commented:
Converting to draft since it is contingent on the decision in https://github.com/kokkos/kokkos/pull/8402

JBludau committed:
adapt minimal cmake

Signed-off-by: Jakob Bludau <bludauj@ornl.gov>

JBludau committed:
bump cmake to 3.25.2 in example

Signed-off-by: Jakob Bludau <bludauj@ornl.gov>

dalg24 commented on example/build_cmake_installed_multilanguage/library.cpp:
```diff
@@ -0,0 +1,24 @@
+#include <iostream>
+void print_cxx() {
+  std::cout << "Hello From C++ library\n";
+  Kokkos::DefaultExecutionSpace().print_configuration(std::cout);
```
Should we use DefaultHostExecutionSpace here?

JBludau replied:
changed in https://github.com/kokkos/kokkos/pull/8167/commits/c998bce3ea0ba28ce0a98ee449f3191e497ff425

dalg24 commented on example/build_cmake_installed_multilanguage/cmake_example.cpp:
```diff
@@ -0,0 +1,67 @@
+
+int main(int argc, char* argv[]) {
+  Kokkos::ScopeGuard guard(argc, argv);
+  Kokkos::DefaultExecutionSpace().print_configuration(std::cout);
```
Should we do Kokkos::print_configuration() here instead of printing from the default exec space?

JBludau replied:
changed in https://github.com/kokkos/kokkos/pull/8167/commits/c998bce3ea0ba28ce0a98ee449f3191e497ff425

dalg24 commented on example/build_cmake_installed_multilanguage/cmake_example.cpp:
```diff
@@ -0,0 +1,67 @@
+  printf("Number of even integers from 0 to %ld\n", n - 1);
+
+  Kokkos::Timer timer;
+  timer.reset();
```
```suggestion
```

JBludau replied:
removed in https://github.com/kokkos/kokkos/pull/8167/commits/c998bce3ea0ba28ce0a98ee449f3191e497ff425

JBludau commented:
Retest this please

masterleinad commented on example/build_cmake_installed_multilanguage/CMakeLists.txt:
```diff
@@ -0,0 +1,40 @@
+
+# Projects can safely mix languages - must have C++ support
+# Kokkos flags will only apply to C++ files
+project(Example CXX Fortran)
```
I don't think we need `Fortran` for all these examples.

JBludau replied:
removed in https://github.com/kokkos/kokkos/pull/8167/commits/2a95419895e349bfd31aa074df0147f345d75be5

masterleinad reviewed (commented):
It seems this is a superset of what COMPILE_AS_CMAKE_LANGUAGE provides right now. Why do we need a new option instead of adding the capability to COMPILE_AS_CMAKE_LANGUAGE?

JBludau commented:
> It seems this is a superset of what COMPILE_AS_CMAKE_LANGUAGE provides right now. 

No, with `COMPILE_AS_CMAKE_LANGUAGE` Kokkos can only be used in a e.g. CUDA language build. With this change, an installed Kokkos does not care if it is used in CXX or e.g. CUDA. As it should be in CMake's ideology. And I did not want to break people that rely on `COMPILE_AS_CMAKE_LANGUAGE` behaving as before.

> Why do we need a new option instead of adding the capability to COMPILE_AS_CMAKE_LANGUAGE?

Because the two are doing different things. `COMPILE_AS_CMAKE_LANGUAGE` compiles Kokkos itself as e.g. CUDA and only sets flags for the CUDA language. With this change we export flags for both e.g. CUDA and CXX, thus downstream apps will get the right flags as we do not care if we are compiled in a e.g CUDA language or as CXX with a compiler that can consume our code.

Does that clarify it?

masterleinad commented:
> Because the two are doing different things. `COMPILE_AS_CMAKE_LANGUAGE` compiles Kokkos itself as e.g. CUDA and only sets flags for the CUDA language. With this change we export flags for both e.g. CUDA and CXX, thus downstream apps will get the right flags as we do not care if we are compiled in a e.g CUDA language or as CXX with a compiler that can consume our code.

Does it really matter much which language is used for compiling `Kokkos`, though (apart form CMake making using CXX more and more difficult)?
Is there any downside for users if we just always set flags for both languages (apart from requiring a higher CMake version which is why would still want a CMake opt-in option)?

JBludau commented:
> Does it really matter much which language is used for compiling `Kokkos`, though (apart form CMake making using CXX more and more difficult)? 
>Is there any downside for users if we just always set flags for both languages (apart from requiring a higher CMake version which is why would still want a CMake opt-in option)?

There might be users that use the different languages to separate things that otherwise would not compile together. Although I agree, that that is them doing something brittle, I would rather have keep language builds and multilanguage flags separate

JBludau committed:
remove fortran from example

Signed-off-by: Jakob Bludau <bludauj@ornl.gov>

dalg24 commented on example/build_cmake_installed_multilanguage/library.hip:
```diff
@@ -0,0 +1,24 @@
+void print_language() {
+  std::cout << "Hello From HIP library\n";
+  Kokkos::DefaultExecutionSpace().print_configuration(std::cout);
+  std::cout << "Goodbye from HIP library" << std::endl;
```
Do we care about flushing the standard output stream?  (not blocking)

JBludau replied:
removed in https://github.com/kokkos/kokkos/pull/8167/commits/c998bce3ea0ba28ce0a98ee449f3191e497ff425

masterleinad reviewed (commented):
My preference is still to add this feature to existing options but apart from the option name this looks good to me.

dalg24 commented on cmake/kokkos_enable_options.cmake:
```diff
@@ -86,12 +86,12 @@ kokkos_enable_option(COMPILER_WARNINGS OFF "Whether to print all compiler warnin
 kokkos_enable_option(COMPILE_AS_CMAKE_LANGUAGE OFF "Whether to use native cmake language support")
-if(Kokkos_ENABLE_COMPILE_AS_CMAKE_LANGUAGE AND Kokkos_ENABLE_CUDA)
+kokkos_enable_option(MULTIPLE_CMAKE_LANGUAGES OFF "Whether to allow Kokkos to be used with multiple CMake languages")
+if((Kokkos_ENABLE_COMPILE_AS_CMAKE_LANGUAGE OR Kokkos_ENABLE_MULTIPLE_CMAKE_LANGUAGES) AND Kokkos_ENABLE_CUDA)
```
What about multiple cmake language and HIP?

JBludau replied:
https://cmake.org/cmake/help/latest/prop_tgt/HIP_STANDARD.html indicates it is no problem with > 3.21. 
It looks like they have forgotten to add C++20 for cuda in earlier versions

masterleinad commented:
Looking at this again, this is basically a combination of the regular behavior and the CMAKE_LANGUAGE behavior. Which only sets flags for C++ files or CUDA/HIP files respectively. In particular, regulalry Kokkos isn't used for CUDA/HIP files and isn't used for C++ files with CMAKE_LANGUAGE allowing to use a different compiler (and Kokkos itself is also compiled internally using CUDA/HIP). With the new option we always use Kokkos wrappers for C++ and CUDA/HIP files. This means that then the `CMAKE_LANGUAGE` option only controls what language Kokkos is compiled in internally but doesn't have any impact on downstream code. Thus, I would keep the new option separate but make it mutually exclusive with the CMAKE_LANGUAGE option.

JBludau committed:
make multilang and compile as language mutually exclusive

Signed-off-by: Jakob Bludau <bludauj@ornl.gov>

JBludau commented:
> Looking at this again, this is basically a combination of the regular behavior and the CMAKE_LANGUAGE behavior. Which only sets flags for C++ files or CUDA/HIP files respectively. In particular, regulalry Kokkos isn't used for CUDA/HIP files and isn't used for C++ files with CMAKE_LANGUAGE allowing to use a different compiler (and Kokkos itself is also compiled internally using CUDA/HIP). With the new option we always use Kokkos wrappers for C++ and CUDA/HIP files. This means that then the `CMAKE_LANGUAGE` option only controls what language Kokkos is compiled in internally but doesn't have any impact on downstream code. Thus, I would keep the new option separate but make it mutually exclusive with the CMAKE_LANGUAGE option.

I made MULTILANGUAGE and CMAKE_LANGUAGE mutually exclusive in https://github.com/kokkos/kokkos/pull/8167/commits/e2344645df5f88f92638f631cade86e4559c6b81

dalg24 reviewed (commented):
(apparently I had pending comments that I never got around to submit...)

tpadioleau commented on example/build_cmake_installed_multilanguage/CMakeLists.txt:
```diff
@@ -0,0 +1,40 @@
+  enable_language(CUDA)
+  set(CMAKE_CUDA_ARCHITECTURES ${Kokkos_CUDA_ARCHITECTURES})
+
+  add_library(language_lib library.cu)
+  target_link_libraries(language_lib PUBLIC Kokkos::kokkos)
```
Just to be sure, as Kokkos is linked to the target `language_lib`, when we enable rdc in Kokkos, is it also propagated to this target ?

JBludau replied:
I will check explicitly and report back but we add them to our flags for both `CXX` and `CUDA` so they should be there

JBludau replied:
```
Building CUDA object CMakeFiles/language_lib.dir/library.cu.o
/kokkos_install/bin/kokkos_launch_compiler /kokkos_install/bin/nvcc_wrapper /usr/bin/c++ /usr/local/cuda/bin/nvcc -forward-unknown-to-host-compiler -DKOKKOS_DEPENDENCE --options-file CMakeFiles/language_lib.dir/includes_CUDA.rsp -std=c++20 "--generate-code=arch=compute_90,code=[compute_90,sm_90]" --relocatable-device-code=true -extended-lambda -Wext-lambda-captures-this -arch=sm_90 -MD -MT CMakeFiles/language_lib.dir/library.cu.o -MF CMakeFiles/language_lib.dir/library.cu.o.d -x cu -c /kokkos/example/build_cmake_installed_multilanguage/library.cu -o CMakeFiles/language_lib.dir/library.cu.o
```
and
```
Building CXX object CMakeFiles/cxx_lib.dir/library.cpp.o
/kokkos_install/bin/kokkos_launch_compiler /kokkos_install/bin/nvcc_wrapper /usr/bin/c++ /usr/bin/c++ -DKOKKOS_DEPENDENCE -isystem /kokkos_install/include -isystem /usr/local/cuda/targets/sbsa-linux/include -std=c++20 --relocatable-device-code=true -extended-lambda -Wext-lambda-captures-this -arch=sm_90 -MD -MT CMakeFiles/cxx_lib.dir/library.cpp.o -MF CMakeFiles/cxx_lib.dir/library.cpp.o.d -o CMakeFiles/cxx_lib.dir/library.cpp.o -c /kokkos/example/build_cmake_installed_multilanguage/library.cpp
```
both have the RDC flag

tpadioleau replied:
Thank you, that is reassuring. I am surprised to see Kokkos mechanisms (`/kokkos_install/bin/kokkos_launch_compiler /kokkos_install/bin/nvcc_wrapper /usr/bin/c++ /usr/local/cuda/bin/nvcc`) on the CUDA target, I would have rather expected a direct call to `nvcc` ?

JBludau replied:
currently we only can set a global launcher with cmake ... They are working on it.
In this case the launcher directly calls nvcc (since the second and third argument are not the same).

JBludau replied:
just saw that the CI also tests the rdc case

tpadioleau commented on cmake/kokkos_tribits.cmake:
```diff
@@ -235,10 +235,20 @@ endmacro()
     #I can use link options
     #check for CXX linkage using the simple 3.18 way
-    target_link_options(${LIBRARY_NAME} PUBLIC $<$<LINK_LANGUAGE:CXX>:${KOKKOS_LINK_OPTIONS}>)
+    target_link_options(${LIBRARY_NAME} PUBLIC $<$<LINK_LANGUAGE:${Kokkos_LANGUAGES}>:${KOKKOS_LINK_OPTIONS}>)
```
Cmake does not require to enable languages before using them in generator expressions ? Or is it enabled somewhere else ?

JBludau replied:
No, the idea is that these generator expressions get carried through the generation stage and only get inserted when the actual build command is executed.

tpadioleau reviewed (commented):
This seems to me a great feature for better interoperability with other libraries.

cedricchevalier19 commented on .jenkins/continuous.groovy:
```diff
@@ -198,6 +199,14 @@ pipeline {
                               .. && \
                               make -j8 && ctest --verbose && \
+                              cd ../../build_cmake_installed_multilanguage && \
+                              rm -rf build && mkdir -p build && cd build && \
+                              cmake \
+                                -DCMAKE_CXX_COMPILER=g++-11 \
+                                -DCMAKE_CXX_FLAGS=-Werror \
+                                -DCMAKE_CXX_STANDARD=20 \
+                              .. && \
+                              make -j8 && ctest --verbose && \
```
Drive-by, but why not using `cmake` for all the phases, including the build?

JBludau replied:
I copied what we already do in our configs. I agree but think this could be done in another pr

cedricchevalier19 commented on example/build_cmake_installed_multilanguage/cmake_example.cpp:
```diff
@@ -0,0 +1,65 @@
+  long seq_count = 0;
+  for (long i = 0; i < n; ++i) {
+    seq_count += (i % 2) == 0;
+  }
```
Not important but do we really need this code in this example? It adds some code that is not related to the feature and a comparison to a pure Cuda or HIP kernel in the linked library looks more interesting as an example.

JBludau replied:
agreed, removed in https://github.com/kokkos/kokkos/pull/8167/commits/c998bce3ea0ba28ce0a98ee449f3191e497ff425

cedricchevalier19 commented on cmake/kokkos_tribits.cmake:
```diff
@@ -235,10 +235,20 @@ endmacro()
     #I can use link options
     #check for CXX linkage using the simple 3.18 way
-    target_link_options(${LIBRARY_NAME} PUBLIC $<$<LINK_LANGUAGE:CXX>:${KOKKOS_LINK_OPTIONS}>)
+    target_link_options(${LIBRARY_NAME} PUBLIC $<$<LINK_LANGUAGE:${Kokkos_LANGUAGES}>:${KOKKOS_LINK_OPTIONS}>)
```
Are `${KOKKOS_LINK_OPTIONS}` always correct for all languages?

JBludau replied:
That is a good question, I will take a deeper look but I would assume they are

JBludau replied:
No, they are not. This means, that using the same Kokkos with e.g. nvcc as `CUDA` compiler and clang as `CXX` compiler might lead to linker problems. But this might lead to ABI problems, too.
This PR would allow using multiple languages with one compatible compiler family.
We could check if the `CUDA/HIP` and `CXX` compiler are compatible but then we would require both languages to be enabled in CMake.

In the long run we can only solve this completely by making the impls we put into `libkokkos.a/so` compiler independent and then use CMake's generator expressions to set the flags for all possible compilers. Then, whatever ABI compatible compilers are used, we get a working build. But that is a massive effort.

cedricchevalier19 replied:
Unfortunately ABI compatibility is terrible with these technologies. However, in the case of using a compiler wrapper I think you are in the case where it can run quite smoothly.
I think your PR is a progress for interoperability so it is worth to merge it in the current state.

cedricchevalier19 reviewed (commented):
It is late and I have to think more about the logic but at first it looks weird to put some Cuda or HIP options to language CXX.

JBludau committed:
simplify example

Signed-off-by: Jakob Bludau <bludauj@ornl.gov>

JBludau committed:
add hip language build

Signed-off-by: Jakob Bludau <bludauj@ornl.gov>

JBludau committed:
use 3.22 for example

Signed-off-by: Jakob Bludau <bludauj@ornl.gov>

tpadioleau reviewed (approved):
Ok for me!

cedricchevalier19 reviewed (approved):
I think this is a great addition for interoperability when using compiler wrappers.

JBludau commented:
@ibaned would this also benefit your use case?

JBludau commented:
The CI failure seems unrelated
